pkg/gcp/secret.go

```diff
@@ -0,0 +1,57 @@
+package gcp
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"time"
+
+	"google.golang.org/api/secretmanager/v1"
+)
+
+// SecretClient is GCP compute client with go client
+type SecretClient struct {
+	service   *secretmanager.Service
+	projectID string
+}
+
+// NewSecretClient returns SecretClient
+func NewSecretClient(
+	ctx context.Context,
+	projectID string,
+) (*SecretClient, error) {
+	s, err := secretmanager.NewService(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SecretClient{
+		service:   s,
+		projectID: projectID,
+	}, nil
+}
+
+func (c *SecretClient) GetSecret(name string) (*secretmanager.Secret, error) {
+	return c.service.Projects.Secrets.Get(fmt.Sprintf("projects/%s/secrets/%s", c.projectID, name)).Do()
+}
+
+func (c *SecretClient) CreateSecretFromData(expireTime time.Time, secretName string, data string) error {
+	encodedData := base64.StdEncoding.EncodeToString([]byte(data))
+	formattedExpireTime := expireTime.Format(time.RFC3339)
+	parent := "projects/" + c.projectID
+	secretPath := fmt.Sprintf("%s/secret/%s", parent, secretName)
+	_, err := c.service.Projects.Secrets.Get(secretPath).Do()
+	if err == nil {
+		return fmt.Errorf("secrets %s already exists", secretPath)
+	}
+	newSecret := &secretmanager.Secret{ExpireTime: formattedExpireTime, Replication: &secretmanager.Replication{Automatic: &secretmanager.Automatic{}}}
+	_, err = c.service.Projects.Secrets.Create(parent, newSecret).SecretId(secretName).Do()
+	if err != nil {
+		return err
+	}
```

Comment:
Can we omit the `Secrets.Get` here?
If the `Secrets.Create` returns an error of already exists, we can delete the get function.

Reply:
The error returned by`Secrets.Create` have only code and Header.
So it would be difficult to confirm that the error means `already exists`.

Reply:
As far as I tried, we can confirm as follows.
However the `alreadyExists` is not documented. So there is a slight risk of changing without notice.
(But I think this is sufficient for our purposes.)

```golang
func IsAlreadyExists(err error) bool {
	if e, ok := err.(*googleapi.Error); ok && e.Code == 409 {
		// Error code 409 is ALREADY_EXISTS or ABORTED. So check the reason as well.
		// However, the string "alreadyExists" is not documented in the official google documentation...
		// See https://cloud.google.com/apis/design/errors#error_mapping
		for _, item := range e.Errors {
			if item.Reason == "alreadyExists" {
				return true
			}
		}
	}
	return false
}
```